cpp/src/text/subword/data_normalizer.cu

```diff
@@ -0,0 +1,459 @@
+
+  if (char_for_thread < total_bytes) {
+    auto const code_point = extract_code_points_from_utf8(strings, char_for_thread);
+    auto const metadata   = cp_metadata[code_point];
+
+    if (is_head_byte(strings[char_for_thread]) && !should_remove_cp(metadata, do_lower_case)) {
+      num_new_chars = 1;
+      // Apply lower cases and accent stripping if necessary
+      auto const new_cp =
+        do_lower_case || always_replace(metadata) ? get_first_cp(metadata) : code_point;
+      replacement_code_points[0] = new_cp == 0 ? code_point : new_cp;
+
+      if (do_lower_case && is_multi_char_transform(metadata)) {
+        auto const next_cps          = aux_table[code_point];
+        replacement_code_points[1]   = static_cast<uint32_t>(next_cps >> 32);
+        auto const potential_next_cp = static_cast<uint32_t>(next_cps);
+        replacement_code_points[2] =
+          potential_next_cp != 0 ? potential_next_cp : replacement_code_points[2];
+        num_new_chars = 2 + (potential_next_cp != 0);
+      }
+
+      if (should_add_spaces(metadata, do_lower_case)) {
+        // Need to shift all existing code-points up one
+        // This is a rotate right. There is no thrust equivalent at this time.
+        for (int loc = num_new_chars; loc > 0; --loc) {
+          replacement_code_points[loc] = replacement_code_points[loc - 1];
+        }
+
+        // Write the required spaces at the end
+        replacement_code_points[0]                 = SPACE_CODE_POINT;
+        replacement_code_points[num_new_chars + 1] = SPACE_CODE_POINT;
+        num_new_chars += 2;
+      }
+    }
+  }
+
+  chars_per_thread[char_for_thread] = num_new_chars;
+
+  typedef cub::
+    BlockStore<uint32_t, THREADS_PER_BLOCK, MAX_NEW_CHARS, cub::BLOCK_STORE_WARP_TRANSPOSE>
+      BlockStore;
+  __shared__ typename BlockStore::TempStorage temp_storage;
+
+  // Now we perform coalesced writes back to global memory using cub.
+  uint32_t* block_base = code_points + blockIdx.x * blockDim.x * MAX_NEW_CHARS;
+  BlockStore(temp_storage).Store(block_base, replacement_code_points);
+}
+
+// There are fixed lookup tables that are copied from codepoint_metadata.ah
+// to the GPU. The per-context-cache ensures there is only one copy per context.
+cudf::strings::detail::thread_safe_per_context_cache<codepoint_metadata_type> g_codepoint_metadata;
+cudf::strings::detail::thread_safe_per_context_cache<aux_codepoint_data_type> g_aux_codepoint_data;
+
+/**
+ * @brief Retrieve the code point metadata table.
+ *
+ * Build the code point metadata tables in device memory
+ * using the vector pieces from codepoint_metadata.ah
+ */
+const codepoint_metadata_type* get_codepoint_metadata(cudaStream_t stream)
+{
+  return g_codepoint_metadata.find_or_initialize([stream](void) {
+    codepoint_metadata_type* table =
+      static_cast<codepoint_metadata_type*>(rmm::mr::get_default_resource()->allocate(
+        codepoint_metadata_size * sizeof(codepoint_metadata_type), stream));
```

Comment:
This will not work since the table will go out of scope at the end of the lambda and free the data. The data needs to live pass the end of the function. It is a global table that is a singleton to be loaded only once per context. I tried making the `uvector` a  function-local static but it seg-faults during process exit.